esphome/components/nfc/ndef_message.cpp

```diff
@@ -38,28 +35,34 @@ NdefMessage::NdefMessage(std::vector<uint8_t> &data) {
     ESP_LOGVV(TAG, "Lengths: type=%d, payload=%d, id=%d", type_length, payload_length, id_length);
 
     std::string type_str(data.begin() + index, data.begin() + index + type_length);
-    record->set_type(type_str);
+
     index += type_length;
 
+    std::string id_str = "";
     if (il) {
-      std::string id_str(data.begin() + index, data.begin() + index + id_length);
-      record->set_id(id_str);
+      id_str = std::string(data.begin() + index, data.begin() + index + id_length);
       index += id_length;
     }
 
-    uint8_t payload_identifier = 0x00;
-    if (type_str == "U") {
-      payload_identifier = data[index++];
-      payload_length -= 1;
-    }
+    std::vector<uint8_t> payload_data(data.begin() + index, data.begin() + index + payload_length);
 
-    std::string payload_str(data.begin() + index, data.begin() + index + payload_length);
+    auto record = make_unique<NdefRecord>();
```

Comment:
Given that the value is immediately overwritten, it doesn't need to be initialized:
```suggestion
    std::unique_ptr<NdefRecord> record;
```